File: tests/fixtures/python/hierarchy.py
"""Multi-level inheritance fixture for `implements` transitive tests.

Covers: abstract base + multiple concrete subclasses + grandchild +
great-grandchild, a multiple-inheritance diamond, and a Protocol-style
interface chain.
"""
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Protocol


class Animal(ABC):
    @abstractmethod
    def eat(self) -> None:
        ...


class Dog(Animal):
    def eat(self) -> None:
        pass


class Cat(Animal):
    def eat(self) -> None:
        pass


# Transitive through Dog.
class Puppy(Dog):
    def play(self) -> None:
        pass


# Two-level transitive: Animal ← Dog ← Puppy ← Pomeranian.
class Pomeranian(Puppy):
    def yap(self) -> None:
        pass


# Protocol chain — transitive via Readable.
class Readable(Protocol):
    def read(self) -> str: ...


class SizedReadable(Readable, Protocol):
    def size(self) -> int: ...


class FileReader(SizedReadable):
    def read(self) -> str:
        return ""

    def size(self) -> int:
        return 0

File: tests/fixtures/java/hierarchy.java
package com.example.demo.hierarchy;

// Multi-level inheritance fixture used by the `implements` transitive
// tests. Three levels (Base → Middle → Leaf), a diamond (both Middle
// variants as bases of Diamond), and a silly self-cycle that tree-sitter
// still parses — the search must not loop forever.

public abstract class Animal {
    public abstract void eat();
}

// Direct subclass of Animal.
public class Dog extends Animal {
    @Override
    public void eat() {}
}

// Direct subclass of Animal — branches away from Dog.
public class Cat extends Animal {
    @Override
    public void eat() {}
}

// Grandchild — transitive subclass of Animal via Dog.
public class Puppy extends Dog {
    public void play() {}
}

// Great-grandchild — transitive via Dog → Puppy.
public class Pomeranian extends Puppy {
    public void yap() {}
}

// Diamond: Mixed has both Dog and Cat as bases — Java single-inheritance
// forbids this for classes, but we approximate with an interface + class.
public interface Mixable {}

public class Mixed extends Dog implements Mixable {}

// Self-referencing cycle (nonsense but syntactically valid). The cycle
// walker must not loop on this.
public class Loopy extends Loopy {}
